pkg/reconciler/oam/workload/translate.go

```diff
@@ -0,0 +1,151 @@
+limitations under the License.
+*/
+
+package workload
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	errInjectService = "unable to inject Service in objects"
+)
+
+var (
+	deploymentGroupVersionKind = appsv1.SchemeGroupVersion.WithKind(reflect.TypeOf(appsv1.Deployment{}).Name())
+
+	serviceKind       = reflect.TypeOf(corev1.Service{}).Name()
+	serviceAPIVersion = corev1.SchemeGroupVersion.String()
+)
+
+var labelKey = "workload.oam.crossplane.io"
+
+// A Translator is responsible for packaging workloads into other objects.
+type Translator interface {
+	Translate(context.Context, resource.Workload) ([]Object, error)
+}
+
+// An ObjectTranslator is a concrete implementation of a Translator.
+type ObjectTranslator struct {
+	TranslateFn
+}
+
+// Translate a workload into other objects.
+func (p *ObjectTranslator) Translate(ctx context.Context, w resource.Workload) ([]Object, error) {
+	return p.TranslateFn(ctx, w)
+}
+
+// NewObjectTranslatorWithWrappers returns a Translator that translates and wraps
+// a workload.
+func NewObjectTranslatorWithWrappers(t TranslateFn, wp ...TranslationWrapper) Translator {
+	return &ObjectTranslator{
+		TranslateFn: func(ctx context.Context, w resource.Workload) ([]Object, error) {
+			objs, err := t(ctx, w)
+			if err != nil {
+				return nil, err
+			}
+			for _, wrap := range wp {
+				if objs, err = wrap(ctx, w, objs); err != nil {
+					return nil, err
+				}
+			}
+			return objs, nil
+		},
+	}
+}
+
+// A TranslateFn translates a workload into an object.
+type TranslateFn func(context.Context, resource.Workload) ([]Object, error)
+
+// Translate workload into object or objects with no wrappers.
+func (fn TranslateFn) Translate(ctx context.Context, w resource.Workload) ([]Object, error) {
+	return fn(ctx, w)
+}
+
+var _ Translator = TranslateFn(NoopTranslate)
+
+// NoopTranslate does not translate the workload and does not return error.
+func NoopTranslate(ctx context.Context, w resource.Workload) ([]Object, error) {
+	return nil, nil
+}
+
+// A TranslationWrapper wraps the output of a workload translation in another
+// object or adds addition object.
+type TranslationWrapper func(context.Context, resource.Workload, []Object) ([]Object, error)
+
+var _ TranslationWrapper = NoopWrapper
+
+// NoopWrapper does not wrap the workload translation and does not return error.
+func NoopWrapper(ctx context.Context, w resource.Workload, objs []Object) ([]Object, error) {
+	return objs, nil
+}
+
+var _ TranslationWrapper = ServiceInjector
+
+// ServiceInjector adds a Service object for the first Port on the first
+// Container for the first Deployment observed in a workload translation.
+func ServiceInjector(ctx context.Context, w resource.Workload, objs []Object) ([]Object, error) {
```

Comment:
@negz I kept this for now since it only relies on k8s types and seems like it could be reused across remote and local OAM addons. Happy to remove if you prefer.

Reply:
I'm not exactly sure where these should live, but I suspect crossplane-runtime is not the right place. Do we have other examples of functions in this Go module that operate on core Kubernetes resource kinds?

Reply:
I don't think we do.. let's just keep them in `addon-oam-kubernetes-remote` for now :+1: